File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/types/Enumeration.java
package me.centralhardware.znatoki.telegram.statistic.eav.types;

import io.vavr.control.Validation;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.util.Arrays;
import java.util.Optional;

public non-sealed class Enumeration implements Type {
    @Override
    public String format(String name, Boolean isOptional) {
        return STR."Выберите \{name}. \{isOptional? optionalText: ""}";
    }

    @Override
    public Validation<String, Void> __validate(Update update, String...variants) {
        return update.hasMessage() && Arrays.asList(variants).contains(update.getMessage().getText())?
                Validation.valid(null) :
                Validation.invalid("Выберите вариант из кастомный клавиатуры");
    }

    @Override
    public Validation<String, Void> validate(String value) {
        return null;
    }

    @Override
    public Optional<String> extract(Update update) {
        return Optional.ofNullable(update.getMessage().getText());
    }

    @Override
    public String getName() {
        return "Enumeration";
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/types/Enumeration.kt
package me.centralhardware.znatoki.telegram.statistic.eav.types

import arrow.core.Either
import dev.inmo.tgbotapi.extensions.utils.extensions.raw.text
import dev.inmo.tgbotapi.types.message.abstracts.CommonMessage
import dev.inmo.tgbotapi.types.message.content.MessageContent
import dev.inmo.tgbotapi.types.message.content.TextContent
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type.Companion.OPTIONAL_TEXT

typealias EnumType = Enumeration
data object Enumeration : Type {
    override fun format(name: String, isOptional: Boolean): String {
        return "Выберите $name. ${if (isOptional) OPTIONAL_TEXT else ""}"
    }

    override fun validate(message: CommonMessage<MessageContent>, variants: List<String>): Either<String, Unit> {
        return if (message.content is TextContent && variants.contains(message.text)) {
            Either.Right(Unit)
        } else {
            Either.Left("Выберите вариант из кастомный клавиатуры")
        }
    }

}